lib/commands/web.js

```diff
@@ -145,37 +129,47 @@ extensions.translateWebCoords = async function (coords) {
 
   webview = util.unwrapElement(webview);
   let rect = await this.proxyCommand(`/element/${webview}/rect`, 'GET');
-  let wvPos = {x: rect.x, y: rect.y};
-  let realDims = {w: rect.width, h: rect.height};
+  log.debug(`Reported webview native rect: ${JSON.stringify(rect)}`);
+
+  // The webview always reports its rect to be the same as the whole app, even though it's not.
+  // There are various elements above and below it. So we have to account for these elements
+  // ourselves.
+
+  let topElementsHeight = await this.getURLBarHeight();
+  topElementsHeight += await this.getHeightOfExtraTopElements();
 
-  let cmd = '(function () { return {w: document.documentElement.clientWidth, h: document.documentElement.clientHeight}; })()';
-  let wvDims = await this.remote.execute(cmd);
+  let bottomElementsHeight = await this.getBottomBarHeight();
+
+  rect.y += topElementsHeight;
+  rect.height -= (topElementsHeight + bottomElementsHeight);
+
+  log.debug(`Corrected webview native rect: ${JSON.stringify(rect)}`);
+  return rect;
+};
+
+extensions.translateWebCoords = async function (coords) {
+  log.debug(`Translating web coordinates (${JSON.stringify(coords)}) to native coordinates`);
 
-  // TODO: investigate where these come from. They appear to be constants in my tests
-  let urlBarHeight = 64;
-  wvPos.y += urlBarHeight;
+  let wvNativeRect = await this.getWebviewNativeRect();
 
-  let realDimensionHeight = 108;
-  realDims.h -= realDimensionHeight;
+  let cmd = '(function () { return {width: window.innerWidth, height: window.innerHeight}; })()';
+  let wvWebDims = await this.remote.execute(cmd);
```

Comment:
This could be `_.memoize`d, with the orientation as a parameter.  From past experience with iOS, each command "over the wire" takes a non-negligible amount of time.  Maybe that's no longer true?

Possibly other dimensions could be memoized similarly, but you mention that the interaction with scrolling is non-trivial, so perhaps there is no suitable parameter to memoize on.  The bottom bar could be memoized on orientation, I think, but the top bar can't?

Reply:
On second thought, the `innerWidth/Height` may change if the `<meta name="viewport" ...>` scale factor thing, or some other zoom parameter, is present in the page.  If that's true then it could only be memoized per page as well as per orientation, which is perhaps too tricky.

Reply:
Oh sod it, memoizing is more effort than it's worth for a prototype, it may just be a solution without a problem.

Reply:
This stuff is indeed pretty slow - a few seconds to do a native tap. So it could do with speeding up, and we may need to work on it at some point. But I agree the prototype doesn't need it.

Memoising this is tricky because the page can be zoomed (happens by default if you open an input widget, for example). And I think the WDA calls are the slow ones, but that's just an impression.